go/service/usersearch.go

```diff
@@ -197,18 +202,51 @@ func (h *UserSearchHandler) UserSearch(ctx context.Context, arg keybase1.UserSea
 		return res, nil
 	}
 
+	if arg.Service != "keybase" && arg.Service != "" {
+		// If this is a social search, we just return API results.
+		return doSearchRequest(mctx, arg)
+	}
+
 	res, err = doSearchRequest(mctx, arg)
+	if err != nil {
+		mctx.Warning("Failed to do an API search for %q: %s", arg.Service, err)
+	}
+
 	if arg.IncludeContacts {
 		contactsRes, err := contactSearch(mctx, arg)
 		if err != nil {
-			return nil, err
-		}
-		if len(contactsRes) > 0 {
-			var res2 []keybase1.APIUserSearchResult
-			res2 = append(res2, contactsRes...)
-			res2 = append(res2, res...)
-			res = res2
+			mctx.Warning("Failed to do contacts search: %s", err)
+		} else {
+			res = append(res, contactsRes...)
+			// Sort first - we are going to be deduplicating on usernames,
+			// entries with higher score have precedence.
+			sort.Slice(res, func(i, j int) bool {
+				return res[i].RawScore > res[j].RawScore
+			})
```

Comment:
If you sort by your deduping metric first (`username`, or `{username, displayLabel}` depending on your answers to my previous questions), then you can get away with not building the set at all. Sort once by deduping metric, then iterate through the list, picking the highest score of any dupes, and then one final sort by score.

Reply:
Alternatively, you can build a heap indexed by deduping metric, only overwriting if the score is higher, then pull each entry into a new heap, now indexed on score. Finally you can pull out `MaxResults` entries from the heap.

Reply:
I really don't want to spend too much time on that file right now. There are more substantial issues in how contact syncing works right now that the way they are fetched here might change anyway.